table leader lease library

vivekmenezes opened the pull request:
Once schema changes are scheduled to be executed through mutations,
a goroutine on the cluster acquires the table leader lease and executes
schema changes on the table. Any goroutine can acquire the lease.
In the future, the goroutine that executes the schema change command
(RENAME, ALTER, etc) will try to acquire the lease immediately
after marking the up_version bit on the table descriptor. Additionally
goroutines on every node will wake up on gossip notifications and will
attempt to acquire the lease if they see outstanding up_version bit
or mutations on a table descriptor.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3339)

<!-- Reviewable:end -->

vivekmenezes commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[kv/db.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uba9Ox8cuGNaWA4-e:-K4uba9Ox8cuGNaWA4-f:-1052545783) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88cc238a0b67e4e21526ea9acbe036bd84eb673/kv/db.go#L106)):*
This file should not be part of this change. Any idea on what I need to add to the kv client to get rid of this?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/db.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uba9Ox8cuGNaWA4-e:-K4ugfaWqUocIKrk316b:2072335008) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88cc238a0b67e4e21526ea9acbe036bd84eb673/kv/db.go#L106)):*
I'm not sure I understand, are you just trying to drop this diff from the PR?

---

*[sql/schema_changer.go, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uh5JSMiH5uCwkoqdK:-K4uh5JSMiH5uCwkoqdL:-2146036732) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5339507e958cca0e0e3b01fec7f3baccedcd568/sql/schema_changer.go#L69)):*
doesn't seem like this needs to be exported.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/db.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uba9Ox8cuGNaWA4-e:-K4wchLFkErovnK9hCNo:678636750) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88cc238a0b67e4e21526ea9acbe036bd84eb673/kv/db.go#L106)):*
Yes. I needed to put it in because of the inclusion of the call to SetSystemDBTrigger() in all the methods. I can see that when methods call SetSystemDBTrigger() through the sql.Executor they don't trigger this. Is there something I need to explicitly set?

---

*[sql/schema_changer.go, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uh5JSMiH5uCwkoqdK:-K4wdAU0YWrWaDJ6lYn_:1917976624) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5339507e958cca0e0e3b01fec7f3baccedcd568/sql/schema_changer.go#L69)):*
This method is called from package sql_test and therefore must be exported.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/db.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uba9Ox8cuGNaWA4-e:-K4xSv8jOYsEd5KCUmki:1866072662) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88cc238a0b67e4e21526ea9acbe036bd84eb673/kv/db.go#L106)):*
I used TestServer.DB() and that fixed it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

bdarnell commented on sql/parser/builtins.go:
```diff
@@ -1300,7 +1300,8 @@ var uniqueBytesState struct {
 }
 
-func generateUniqueBytes(nodeID uint32) DBytes {
+// GenerateUniqueBytes generates unique bytes for a unique id.
```
Now that this is exported the comment should explain what properties the generated value has (i.e. why isn't it a random UUID? It looks like the major goal is a rough sortability, which doesn't matter for this change so I'm not sure this is the right function to use here). We should also clean up the interface: return `[]byte` instead of the sql-specific `DBytes` and the argument should probably be a `roachpb.NodeID`.

bdarnell commented on sql/schema_changer.go:
```diff
@@ -54,3 +59,77 @@ func (p *planner) applyMutations(tableDesc *TableDescriptor) error {
+	}
+
+	now := time.Now()
+	// Add 5 seconds to the lease expiration to deal with time uncertainty across nodes.
```
The 5 seconds should perhaps be a multiple of the configured MaxOffset. Is there a way to use database time (via a `Deadline` on all `EndTransactionRequests` issued under the authority of this lease) to avoid any question of clock offsets?

bdarnell commented on sql/schema_changer.go:
```diff
@@ -54,3 +59,77 @@ func (p *planner) applyMutations(tableDesc *TableDescriptor) error {
+		return nil, util.Errorf("no lease present for tableID: %d", tableID)
+	}
+	if tableDesc.Lease.Leader != lease.Leader {
+		return nil, util.Errorf("lease for table ID: %d, owned by different leader: %q, expected: %q", tableID, tableDesc.Lease.Leader, lease.Leader)
```
The strings in `Lease.Leader` are opaque, so it doesn't make much sense to print them. I think it would be good for debugability to store the leader's node ID explicitly in the lease (and then a lease instance should be uniquely identifiable by the node ID and expiration time)

bdarnell commented on sql/structured.proto:
```diff
@@ -177,6 +177,16 @@ message TableDescriptor {
+  // The table leader lease. A single goroutine across a cockroach cluster
+  // can own it and as a result will execute scheduled schema changes for
+  // this table. 
+  message LeaderLease {
```
Can we give this a more descriptive name than `LeaderLease`? The word "leader" is already overloaded between raft leadership and the `storage` range leader lease. Maybe `SchemaLease`?

tamird commented:
Can you make the commit message "summary" line more descriptive? Maybe "sql: implement schema lease mechanics"

---

Reviewed 3 of 8 files at r1, 2 of 2 files at r2, 2 of 3 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/schema_changer.go, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uh5JSMiH5uCwkoqdK:-K4xsfEngO_Vh_0lDZ_t:434958769) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5339507e958cca0e0e3b01fec7f3baccedcd568/sql/schema_changer.go#L69)):*
Looks like it's only called from sql_test; can you put it in that package instead of here?

---

*[sql/schema_changer.go, line 65 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtS12M47KhWCNgGhq:-K4xtS12M47KhWCNgGhr:-332571837) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L65)):*
how come pointer? what other structures share this planner?

---

*[sql/schema_changer.go, line 77 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtXpjunP8srSW3kZE:-K4xtXpjunP8srSW3kZF:1978766704) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L77)):*
this method should have a pointer receiver.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-:-K4xsQ4D1B91nHb0bJJF:-1717879780)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
done

---

Review status: 0 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/parser/builtins.go, line 1303 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uys-r3-1303:-K4ycVzKNCdCXPK2mqwi:-1393757274) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/parser/builtins.go#L1303)):*
I've cleaned up the interface for use here. I think it's okay for us to use it here to generate a unique ID.

---

*[sql/schema_changer.go, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4uh5JSMiH5uCwkoqdK:-K4yKvJXLB5mLjTH7GDx:-1678186152) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5339507e958cca0e0e3b01fec7f3baccedcd568/sql/schema_changer.go#L69)):*
I don't think so. This was created to keep planner, sql internal.

---

*[sql/schema_changer.go, line 65 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtS12M47KhWCNgGhq:-K4yJOsoqeCOK5R-uhXQ:1805017525) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L65)):*
This is meant to be called from either the executor or a planner method. Since the planner is  usually referred to using a pointer I've done the same here.

---

*[sql/schema_changer.go, line 77 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtXpjunP8srSW3kZE:-K4yLD98i_7TCrNsZ8Fo:-1964683657) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L77)):*
Any particular reason for that?

---

*[sql/schema_changer.go, line 85 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyt-r3-85:-K4yLaJE7_LcVkKTzfLY:-1329333703) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L85)):*
Well, this lease is not a strict single writer lease. Anyone holding this lease will move  the execution of the schema change forward via transactions on the DB, and since the execution is via transactions, it is perfectly valid for two or more goroutines to think they are holding a valid lease and execute. So in truth this lease is about reducing write contention by picking one lease holder with very high probability. I've added a comment about this in structured.proto.

---

*[sql/schema_changer.go, line 107 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyt-r3-107:-K4yMdEemYWxEWmi7HwU:-148053325) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L107)):*
I've added the nodeid to the lease to help in debugging. I've kept the Leader around and renamed it to ID.

---

*[sql/structured.proto, line 183 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyv-r3-183:-K4yN2Si2M555MykpFWv:-1795221600) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/structured.proto#L183)):*
named it SchemaChangeLease

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-:-K4yf5vEG8LNRwV8zAkz:677588487)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Ignore the first commit. It's being reviewed in another PR.

---

Review status: 0 of 14 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-:-K4yf8bKiJUh21VaJzrw:-573140220)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 8 files at r1, 2 of 2 files at r2, 2 of 3 files at r3, 10 of 12 files at r4.
Review status: 12 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/schema_changer.go, line 77 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtXpjunP8srSW3kZE:-K4ykiUK2CTDnvXPa8P1:391560861) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L77)):*
it's modifying its receiver. it only works now because `p` happens to be a pointer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 12 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/schema_changer.go, line 77 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtXpjunP8srSW3kZE:-K4ytrgQL8qklyS98akA:1006089967) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L77)):*
Sounds good. Thanks

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r5.
Review status: 12 of 14 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/schema_changer.go, line 65 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtS12M47KhWCNgGhq:-K51St7S_jpX1pyNrbVi:174636844) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L65)):*
That doesn't seem particularly convincing.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 4 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/schema_changer.go, line 65 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4xtS12M47KhWCNgGhq:-K51y_mXQoy9RBSTcoxA:175050851) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L65)):*
I've removed the use of the planner out of this PR. It occurred to me that the sql planner had nothing to do with leases.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 5 files at r6.
Review status: 8 of 9 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

bdarnell commented on sql/schema_changer.go:
```diff
@@ -62,3 +68,81 @@ func (p *planner) applyUpVersion(tableDesc *TableDescriptor) {
+}
+
+// NewSchemaChangerForTesting only for tests.
+func NewSchemaChangerForTesting(tableID ID, node roachpb.NodeID, txn *client.Txn) SchemaChanger {
```
How will a SchemaChanger be created outside of tests? The design seems strange to me: we won't have multiple lease operations within a single transaction, will we? I think the `*client.Txn` should be an argument to the lease methods instead of a field of the `SchemaChanger` (and then what's the point of the SchemaChanger type vs a bunch of static functions?)

bdarnell commented:
Review status: 8 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/schema_changer.go, line 85 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyt-r3-85:-K52GXUqJUVeHqfCN_j_:-1227219721) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L85)):*
A comment would be helpful here, too. And the `5` should still be pulled out into a constant for clarity, or derived from `MaxOffset`.

---

*[sql/schema_changer.go, line 107 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyt-r3-107:-K52FpJ2vPDPVZlSNHhe:875262965) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L107)):*
"ID" is unclear: what does it identify? I think "Token" would be better to make it clear that this is an opaque blob with no other use, but I'm still not clear on why it's even needed (especially with the other comment that the lock is just advisory)

---

*[sql/structured.proto, line 238 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K52H3IfAASK3g-SYDt3:-K52H3IfAASK3g-SYDt4:1135947821) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2cb1b27e94d72b84672753dca66369170464184/sql/structured.proto#L238)):*
Name this `node_id` and give it the `roachpb.NodeID` custom type.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 7 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/schema_changer.go, line 80 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyt-r6-80:-K52LgEVn5K2wqCVgmk1:93677193) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2cb1b27e94d72b84672753dca66369170464184/sql/schema_changer.go#L80)):*
I've modified the schema changer to use client.DB instead of client.Txn. This ensures that each of these methods use their own transaction. The SchemaChanger type was originally SchemaChangeLeaser until I realized there was more in future changes pertaining to the overall rollout of a schema change that I could place in here, such as the mutation ID of the current mutation being operated on.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: implement schema change lease

Once schema changes are scheduled to be executed through mutations,
a goroutine on the cluster acquires the schema change lease and executes
schema changes on the table. Any goroutine can acquire the lease.
In the future, the goroutine that executes the schema change command
(RENAME, ALTER, etc) will try to acquire the lease immediately
after scheduling a schema change  on the table descriptor.
Additionally, an assigned goroutine on every node will wake up on
a gossip notifications and attempt to acquire the lease if it
sees outstanding up_version bit or mutations on a table descriptor.

vivekmenezes commented:
Review status: 4 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/schema_changer.go, line 85 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyt-r3-85:-K52WeEJlF7y12DZ1sBD:-2027964098) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L85)):*
Added a comment and pulled out the constant for clarity.

---

*[sql/schema_changer.go, line 107 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K4ubW0b25FfKb6k1uyt-r3-107:-K52MgX60m_gTlSxoqCS:1892231833) ([raw file](https://github.com/cockroachdb/cockroach/blob/a573e504ac7647cdd2e27ad92e70a326c3eee8a3/sql/schema_changer.go#L107)):*
Yeah, I should just take it out!

---

*[sql/structured.proto, line 238 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339#-K52H3IfAASK3g-SYDt3:-K52U9SenmC267wd5GWd:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2cb1b27e94d72b84672753dca66369170464184/sql/structured.proto#L238)):*
done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 5 files at r8.
Review status: 8 of 9 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3339)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM
